src/clientlayers/StreamRequest.jl

```diff
@@ -103,13 +104,41 @@ writechunk(stream, body::Union{Dict, NamedTuple}) = writebodystream(stream, body
 writechunk(stream, body) = write(stream, body)
 
 function readbody(stream::Stream, res::Response, decompress)
-    readstream = decompress && header(res, "Content-Encoding") == "gzip" ? GzipDecompressorStream(stream) : stream
-    if isbytes(res.body)
-        res.body = read(readstream)
-    elseif !isredirect(stream) && !retryable(stream)
-        # if the request/response pair are going to be redirected or retried,
-        # we want to avoid "contaminating" our response body stream
-        write(res.body, readstream)
+    # Bail early if we are not going to read anything.
+    # If the request/response pair are going to be redirected or retried,
+    # we want to avoid "contaminating" our response body stream.
+    willread = isbytes(res.body) || (!isredirect(stream) && !retryable(stream))
+    willread || return
+
+    if decompress && header(res, "Content-Encoding") == "gzip"
+        # Plug in a buffer stream in between so that we can (i) read the http stream in
+        # chunks instead of byte-by-byte and (ii) make sure to stop reading the http stream
+        # at eof.
+        buf = BufferStream()
+        gzstream = GzipDecompressorStream(buf)
+        tsk = @async begin
```

Comment:
Can you explain why the `@async` + `wait` later? Like, why can't do something like `write(res.body, GzipDecompressorStream(BufferStream(stream)))`?

Reply:
There is  no method `BufferStream(stream)`.

The task is needed since we otherwise have two blocking calls (`write` to the decompressor stream and `read` from the buffer) and we can't make progress. Perhaps there is a better way, but the code in this patch is how I have done things manually when passing `response_stream` before this was built into HTTP.jl at least.